uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+  let v3pool_result = get_get_last_snapshot_value_in_v3pool (targets . get (2usize) . unwrap () . clone ()) . await ;
+  let eth_usdc_price_result = get_get_last_snapshot_value_in_eth_usdc_price (targets . get (3usize) . unwrap () . clone ()) . await ;
+  
+  let fees_24h_usd = pool_fees_result.clone().unwrap().result[0].fees_24h_usd;
+  let t0_decimals = pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t0_decimals;
+  let t1_decimals =  pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t1_decimals;
+
+  let eth_price_usd  = eth_usdc_price_result.clone().unwrap().result.t1_price_usd;
+
+  let address = v3pool_result.clone().unwrap().result.address;
+  let current_tick_liquidity_str = v3pool_result.clone().unwrap().result.liquidity;
+  let current_tick_liquidity = i128::from_str_radix(current_tick_liquidity_str.trim_start_matches("0x"), 16).unwrap();
+  let sqrt_ratio_x96_str = v3pool_result.clone().unwrap().result.sqrt_ratio_x96;
+  let sqrt_ratio_x96 = u128::from_str_radix(sqrt_ratio_x96_str.trim_start_matches("0x"), 16).unwrap();
+  let tick_current = v3pool_result.clone().unwrap().result.tick_current;
+  let tick_spacing = v3pool_result.clone().unwrap().result.tick_spacing;
+  let token0 = v3pool_result.clone().unwrap().result.token0;
+
+  let tick_cumul_28x6h = tc28x6_result.unwrap().0;
+
+  let current_sqrt_price = sqrt_ratio_x96 as f32 / f32::powf(2.0,96.0);
+  let mut current_price = current_sqrt_price * current_sqrt_price;
+
+  let mut compressed: i32 = tick_current / tick_spacing;
+  if tick_current <0 && tick_current % tick_spacing != 0 {
+    compressed -= 1;
+  }
+  let floor = compressed * tick_spacing;
+  let mut min_tick = floor;
+  let mut max_tick = floor + tick_spacing;
+  for i in (1..29).rev() {
+    let new_tick = ((tick_cumul_28x6h[i] - tick_cumul_28x6h[i-1]) / 21600) as i32;
+    if new_tick < min_tick {
+      min_tick = new_tick;
+    } else if new_tick > max_tick {
+      max_tick = new_tick;
+    }
+  }
+  
+  let mut range_top = f32::powf(1.0001, max_tick as f32);
+  let mut range_bottom = f32::powf(1.0001, min_tick as f32);
+  let amount0 = current_tick_liquidity as f32
+    * (range_top.sqrt() - current_sqrt_price as f32)
+    / (current_sqrt_price as f32 * range_top.sqrt())
+    / f32::powf(10.0, t0_decimals as f32);
+  let amount1 = current_tick_liquidity as f32
+    * (current_sqrt_price as f32 - range_bottom.sqrt())
+    / f32::powf(10.0, t1_decimals as f32);
+  let mut est_tvl_in_range = amount0 * current_price + amount1;
+
+  if token0 == "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2" {
+    current_price = 1.0 / current_price;
+    est_tvl_in_range = amount0 + amount1 * current_price;
+    let new_range_top = 1.0 / range_bottom;
+    let new_range_bottom = 1.0 / range_top;
+    range_top = new_range_top;
+    range_bottom = new_range_bottom;
+  }
+  
+  let fees_24h_eth = fees_24h_usd / eth_price_usd;
+  let edr = fees_24h_eth / est_tvl_in_range as f32;
+  let edpr = edr * 100.0;
```

Comment:
I have a project that wants to calculate EDPR, but since the return is a decimal (float), I think it is better to leave it as EDR. The percentages are just a format conversion for the user, and as long as they are numerical, I thought it would be better to leave them as decimals, even if you want to get the results of the Lens calculation and handle them in another canister or in a propagated contract. Also, it is a rare case that a smart contract has data in percentages.

Reply:
https://github.com/horizonx-tech/chainsight-showcase/pull/34#issuecomment-1890472747

Reply:
I still think that canister should be a small number since it is a smart contract to begin with. In many blockchains, it is the front end that should care about this user-oriented notation. And in any case, the computed data will have many zeros, even if it is a percentage. If you care more about the user's perspective, I think current_price should be converted to dollars. If you are not able to use user-oriented notation for everything, I think it would be easier to make decisions with smart contract standards in mind.

Reply:
In conclusion, please use EDR.
The reason is as written in the telegram.

>- EDPR or EDR?
>I don't think it makes much difference since it is two digits that change whether it is a percentage or not. On top of that, I still think decimals are appropriate since we are talking about within the blockchain after all.

Also, algorithm_lens does not need to worry about relayer, and as mentioned earlier, the digits only change by 2 digits.

Reply:
I disagree, also for the reasons given in Telegram. I think that ordinary users (liquidity providers) would find a percentage much more user-friendly and in keeping with other information sources, and a smart contract developer would find it easy to adapt (which they have to do in either case, as the Relayer can only output an integer).

However, I've adjusted the metadata in a way that will hopefully reach a satisfactory middle ground by including EDR, although it makes for a rather long description. I do think that the caveat still needs to be in there, and would suggest that you ask your team's regulatory and legal advisors what they reckon about this.

Not willing to change further, unless you want it reverted to how it was before.

(Note that I'm only saying "not willing" because that's the wording you've asked for. To me it sounds much harsher than the meaning I'm intending to convey.)